modal_images/token_identity.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+from pathlib import Path
+
+from modal import Secret, Stub, gpu, method
+
+from modal_images.gemma import image as gemma_image
+from modal_images.mistral import image as mistral_image
+from obvs.vis import create_heatmap, plot_surprisal
+
+images = {
+    "gemma2": gemma_image,
+    "gemma7": gemma_image,
+    "mistral": mistral_image,
+    "gpt2": gemma_image,
+}
+
+
+model_names = {
+    "llamatiny": "TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T",
+    "llama": "meta-llama/Llama-2-13b-hf",
+    "gpt2": "gpt2",
+    "mamba": "MrGonao/delphi-mamba-100k",
+    "mistral": "mistralai/Mistral-7B-v0.1",
+    "gptj": "EleutherAI/gpt-j-6B",
+    "gemma2": "google/gemma-2b",
+    "gemma7": "google/gemma-7b",
+}
+
+
+stub = Stub(
+    image=gemma_image,
+    name="token_identity",
+    secrets=[Secret.from_name("my-huggingface-secret")],
+)
+
+
+@stub.cls(
+    # gpu=gpu.A100(memory=40, count=1),
+    # gpu=gpu.A10G(count=1),   # 24 GB
+    gpu=gpu.T4(count=1),  # 16 GB
+    # cpu=1,
+    timeout=60 * 30,
+    container_idle_timeout=60 * 5,
+)
+class Runner:
+    def setup_ti(self, model_name):
+        from obvs.lenses import TokenIdentity
+
+        self.ti = TokenIdentity("", model_name)
+
+    @method()
+    def run(self, model_name, prompt, full):
+        if not hasattr(self, "ti"):
+            print("Setting up TokenIdentity")
+            self.setup_ti(model_name)
+        self.ti.patchscope.source.prompt = prompt
+        source_layers = range(self.ti.patchscope.n_layers_source)
+        target_layers = range(self.ti.patchscope.n_layers_target)
+        self.ti.run(
+            source_layers=source_layers,
+            target_layers=target_layers if full else None,
+        ).compute_surprisal()
+        return self.ti.surprisal, self.ti.source_layers
+
+
+@stub.local_entrypoint()
+def main(model_name, n_samples=5, full=False):
+    n_samples = int(n_samples)
+    full = bool(full)
+    import os
+
+    from datasets import load_dataset
+
+    token = os.environ["HUGGINGFACE_TOKEN"].strip()
+    dataset = load_dataset(
+        "oscar-corpus/OSCAR-2201",
+        "en",
+        split="train",
+        streaming=True,
+        token=token,
+    )
+    shuffled_dataset = dataset.shuffle(seed=42, buffer_size=n_samples)
+
+    samples = []
+    for example in shuffled_dataset.take(n_samples):
+        samples.append(example["text"])
+
+    # Trim the samples to the first 300 characters
+    samples = [sample[:1000] for sample in samples]
+
+    # Make sure it ends on a space
+    samples = [sample[: sample.rfind(" ")] for sample in samples]
+
+    # Strip the spaces
+    samples = [sample.strip() for sample in samples]
+
+    filename = Path(f"{model_name}_surprisal_{n_samples}_full").expanduser()
+
+    surprisals = []
+    runner = Runner()
+    for prompt in samples:
+        try:
+            surprisal, layers = runner.run.remote(model_names[model_name], prompt, full)
```

Comment:
in this script, we are sending a request for a token identity run for each sample, and only getting back one or two numbers that we need for graphing. if you ask it to send back the entire output and process it locally, it will eat your bandwidth. we could also move this whole loop into the runner and just get back the final data for vis.

(we could even create the fig objects in the cloud and just get them sent back, but theres not much benefit to that)